Add in-place version of canonical() and more tests

rdeits opened the pull request:
Fixes #429 

This implements the in-place canonicalization I suggested over in #429 with some cleanup to make it a bit more readable. It adds two new functions: `iscanonical()` which tests if an AffineFunction is canonical and `canonical!` which converts an AffineFunction to canonical form in-place. ~~In Julia v0.7 and above, the in-place `canonical!` should allocate zero memory (on v0.6, it allocates a tiny amount for keyword arguments).~~ Edit: the in-place version now has zero allocations on 0.6 and 0.7 (thanks @tkoolen!) 

Another significant change is that if a function is already in canonical form, it is not copied or modified. This should make things a bit faster for modeling layers like JuMP that already provide canonical functions.

rdeits commented:
@tkoolen I think the in-place `MOI.Utilities.canonical!` should work well for SimpleQP, without you needing to keep a `Dict` around.

mohamed82008 commented:
Touches same code as https://github.com/JuliaOpt/MathOptInterface.jl/pull/420.

mlubin commented on src/Utilities/functions.jl:
```diff
@@ -135,45 +165,80 @@ If `x` (resp. `y`, `z`) is `VariableIndex(1)` (resp. 2, 3).
+    f
+end
+
+function merge_duplicates!(x::AbstractVector, by, keep, combine)
```
Where is this function coming from? (It looks like it's copied from Base julia.) It also needs a docstring.

tkoolen replied:
`merge_duplicates!` -> `combine_elements!` / `combine_terms!`?

tkoolen replied:
Or `compress!` / `sort_and_compress!`?

rdeits replied:
@mlubin It came from...me writing it yesterday? Sure, I can add a docstring

rdeits replied:
`sort_and_compress!` seems very clear

mlubin replied:
Sorry, it was a bit early in the morning, didn't mean for it to sound like an accusation if that was the case. I meant that it looks like code that could appear in Base Julia :)

mlubin commented on test/functions.jl:
```diff
@@ -204,4 +204,103 @@
+        srand(1)
+        @testset "ScalarAffineFunction" begin
+            for i in 1:1000
+                f = random_scalar_affine_function(rand(1:1000))
```
It's better to test with interesting corner cases with known correct solutions rather than random input comparing with the old version of the code.

This setup has readability issues because it requires a bunch of work to reproduce and understand what happened when a test failed. It doesn't print the offending function and the observed and expected results. The old reference code is already complicated enough that I can't eyeball to verify that it's correct, so a reader shouldn't have confidence that this test is checking what it claims to test.

rdeits replied:
Sure, we can add explicit test cases and even delete this if you want, but it's been immensely helpful in ensuring that I don't break the implementation while I rewrite it.

tkoolen commented on src/Utilities/functions.jl:
```diff
@@ -135,45 +165,80 @@ If `x` (resp. `y`, `z`) is `VariableIndex(1)` (resp. 2, 3).
+    elseif length(x) > 1
+        sort!(x, QuickSort, Base.Order.ord(isless, by, false, Base.Sort.Forward))
+        i1 = 1
+        i2 = 2
```
Assumes `1:n` indexing, while `x::AbstractVector`.

rdeits replied:
fixed

tkoolen commented on src/Utilities/functions.jl:
```diff
@@ -135,45 +165,80 @@ If `x` (resp. `y`, `z`) is `VariableIndex(1)` (resp. 2, 3).
+        sort!(x, QuickSort, Base.Order.ord(isless, by, false, Base.Sort.Forward))
+        i1 = 1
+        i2 = 2
+        @inbounds while i2 <= length(x)
```
`i2` gets incremented in both branches, so I'd just use `for i2 in eachindex(x)[2 : end]`.

rdeits replied:
fixed

tkoolen commented on src/Utilities/functions.jl:
```diff
@@ -121,6 +121,36 @@ function Base.getindex(it::ScalarFunctionIterator{MOI.VectorAffineFunction{T}},
+                                                       t -> !iszero(t.scalar_term.coefficient))
+iscanonical(f::MOI.ScalarAffineFunction) = iscanonical(f.terms,
+                                                       t -> t.variable_index.value,
+                                                       t -> !iszero(t.coefficient))
```
What about `ScalarQuadraticFunction`, `VectorQuadraticFunction`?

mohamed82008 replied:
Can use the new `termindices` and `coefficient` functions to make this generic.

rdeits replied:
I don't know what we'd consider canonical form for them, as it was never supported before.

rdeits commented on src/Utilities/functions.jl:
```diff
@@ -135,45 +165,80 @@ If `x` (resp. `y`, `z`) is `VariableIndex(1)` (resp. 2, 3).
+Convert a function to canonical form in-place, without allocating a copy to hold the result.
+See [`canonical`](@ref).
+"""
+function canonical!(f::MOI.ScalarAffineFunction)
```
sure

mohamed82008 commented on src/Utilities/functions.jl:
```diff
@@ -135,45 +165,80 @@ If `x` (resp. `y`, `z`) is `VariableIndex(1)` (resp. 2, 3).
+    if iscanonical(f)
+        f
+    else
+        canonical!(MOI.VectorAffineFunction(copy(f.terms), f.constants))
```
Should probably copy `f.constants` too? This part was there before, but might as well change it.

rdeits replied:
Yeah, I was on the fence about that. The constants are never modified during canonicalization, but I guess they might be changed by something downstream. 

I could just implement: 

```
Base.copy(f::ScalarAffineFunction) = <copy terms and constant>
Base.copy(f::VectorAffineFunction) = <copy terms and constants>
```

and then re-write `canonical(f) = canonicalize!(copy(f))` which would work for both

mohamed82008 commented on src/Utilities/functions.jl:
```diff
@@ -135,45 +165,80 @@ If `x` (resp. `y`, `z`) is `VariableIndex(1)` (resp. 2, 3).
+                                                                            t1.scalar_term.variable_index)))
+    f
+end
+
```
Using `unsafe_add`, `coefficient` and `termindices` functions, these 2 `canonical!`s can be combined.

rdeits replied:
Yup, I just saw those from your PR and rewrote the function to use them :)

rdeits commented on src/Utilities/functions.jl:
```diff
@@ -121,6 +121,36 @@ function Base.getindex(it::ScalarFunctionIterator{MOI.VectorAffineFunction{T}},
+iscanonical(f::MOI.ScalarAffineFunction) = iscanonical(f.terms,
+                                                       t -> t.variable_index.value,
+                                                       t -> !iszero(t.coefficient))
+function iscanonical(x::AbstractVector, by, keep)
```
In OSQP you can just do 

```julia
MOI.Utilities.sort_and_compress!(f.terms, MOI.Utilities.termindices, t -> true, MOI.Utilities.unsafe_add)
```

mohamed82008 commented on src/Utilities/functions.jl:
```diff
@@ -241,65 +246,51 @@ function canonical(f::MOI.ScalarAffineFunction)
         f
     else
-        canonical!(MOI.VectorAffineFunction(copy(f.terms), f.constants))
+        canonicalize!(MOI.VectorAffineFunction(copy(f.terms), f.constants))
```
Can use the `_constant` function to merge this and the previous `canonical`s.

rdeits replied:
done (I'm using it in my implementation of Base.copy)

rdeits commented on src/Utilities/functions.jl:
```diff
@@ -121,6 +121,36 @@ function Base.getindex(it::ScalarFunctionIterator{MOI.VectorAffineFunction{T}},
+    if isempty(x)
+        return true
+    end
+    @inbounds for i in 1:(length(x) - 1)
```
fixed

mohamed82008 commented on src/Utilities/functions.jl:
```diff
@@ -242,29 +250,15 @@ If `x` (resp. `y`, `z`) is `VariableIndex(1)` (resp. 2, 3).
-        canonicalize!(MOI.VectorAffineFunction(copy(f.terms), f.constants))
-    end
-end
+canonical(f::Union{SAF, VAF}) = canonicalize!(copy(f))
```
`deepcopy`?

mohamed82008 replied:
Nevermind.

rdeits commented:
@mlubin I've updated the tests with some useful potential corner cases. I kept some limited randomized testing, but I removed the complicated reference implementation, so I'm just testing the invariant that `iscanonical(canonical(f))` is always true in the random cases.

rdeits commented:
Just snuck in a few more tests to try to get the coverage up

mlubin commented on src/Utilities/functions.jl:
```diff
@@ -202,7 +202,42 @@ coefficient(t::MOI.ScalarAffineTerm) = t.coefficient
+function iscanonical(f::Union{SAF, VAF})
+    iscanonical(f.terms, termindices, t -> !iszero(coefficient(t)))
+end
+function iscanonical(x::AbstractVector, by, keep)
```
It's not obvious from the signature the method does, could you add a docstring or (more informally) a comment?

mlubin commented on test/functions.jl:
```diff
@@ -156,6 +156,7 @@
             @test f.terms == MOI.ScalarAffineTerm.([2, 3], [x, y])
             f = MOIU.modifyfunction(f, MOI.ScalarCoefficientChange(x, 0))
             @test f.terms == MOI.ScalarAffineTerm.([3], [y])
+
```
Unnecessary whitespace change?

rdeits replied:
fixed

mlubin commented on test/functions.jl:
```diff
@@ -204,4 +205,149 @@
         end
     end
+    @testset "Conversion to canonical form" begin
+        function _isapprox(f1::T, f2::T) where {T <: Union{MOI.ScalarAffineFunction, MOI.VectorAffineFunction}}
```
The difference between `_isapprox` and `isapprox` ([here](https://github.com/JuliaOpt/MathOptInterface.jl/blob/ae87890bec8922c21df4c817b4e847c67b402240/src/functions.jl#L310)) is confusing and not well explained by an underscore. Maybe call this `isapprox_ordered`?

rdeits replied:
yup, done

mlubin commented on test/functions.jl:
```diff
@@ -204,4 +205,149 @@
+             (MOIU.coefficient.(f1.terms) ≈ MOIU.coefficient.(f2.terms)))
+        end
+        function test_canonicalization(f::T, expected::T) where {T <: Union{MOI.ScalarAffineFunction, MOI.VectorAffineFunction}}
+            @test MOI.Utilities.iscanonical(expected)
```
Inconsistent use of `MOIU` and `MOI.Utilities`.

rdeits replied:
fixed

mlubin commented on test/functions.jl:
```diff
@@ -204,4 +205,149 @@
+                MOI.ScalarAffineFunction(MOI.ScalarAffineTerm.(coefficients, indices), randn())
+            end
+            srand(1)
+            for i in 1:1000
```
I see the potential usefulness of the randomized fuzzing, but we don't use this testing style of testing anywhere else in MOI. It may be confusing to someone reading the code why we would do this to test only `iscanonical`. I would lean towards dropping it but could be swayed otherwise.

rdeits replied:
Ok, removed

mlubin commented on src/Utilities/functions.jl:
```diff
@@ -202,7 +202,49 @@ coefficient(t::MOI.ScalarAffineTerm) = t.coefficient
+"""
+    is_strictly_sorted(x::AbstractVector, by, filter)
+
+Returns `true` if `by(x[i]) < by(x[i + 1]` and `filter(x[i]) == true` for
```
Typo: `by(x[i]) < by(x[i + 1])`

rdeits replied:
fixed, thanks

tkoolen commented on src/Utilities/functions.jl:
```diff
@@ -215,27 +250,52 @@ If `x` (resp. `y`, `z`) is `VariableIndex(1)` (resp. 2, 3).
+"""
+function sort_and_compress!(x::AbstractVector, by, keep, combine)
+    if length(x) > 0
+        sort!(x, QuickSort, Base.Order.ord(isless, by, false, Base.Sort.Forward))
```
Quicksort is in-place, fast, and doesn't allocate. I briefly thought about using counting sort, but that's an out-of-place algorithm, so either you need to preallocate memory somewhere (a hassle) or not care about allocating (which makes this a lot less useful for us and probably also negates any performance gains compared to quicksort in practice).

tkoolen commented on src/Utilities/functions.jl:
```diff
@@ -215,27 +250,52 @@ If `x` (resp. `y`, `z`) is `VariableIndex(1)` (resp. 2, 3).
+    if length(x) > 0
+        sort!(x, QuickSort, Base.Order.ord(isless, by, false, Base.Sort.Forward))
+        i1 = firstindex(x)
+        for i2 in eachindex(x)[2:end]
```
Yeah, it's a generalization. The implementation is different though:

```julia
function unique(f::Callable, C)
    out = Vector{eltype(C)}()
    seen = Set()
    for x in C
        y = f(x)
        if !in(y, seen)
            push!(seen, y)
            push!(out, x)
        end
    end
    out
end
```

(on 0.6). On 0.7 there is also a `unique!(C)` (but interestingly there is no `unique!(f, C)`). And they still allocate and use a different algorithm.

rdeits replied:
It's actually different anyway, as `unique` preserves the original order of appearance rather than sorting. So this is similar to `unique!(sort!(x))`, but I really don't think we need to go into that.

rdeits commented on src/Utilities/functions.jl:
```diff
@@ -215,27 +250,52 @@ If `x` (resp. `y`, `z`) is `VariableIndex(1)` (resp. 2, 3).
+"""
+function canonicalize!(f::Union{SAF, VAF})
+    sort_and_compress!(f.terms, termindices, t -> !iszero(coefficient(t)), unsafe_add)
+    f
```
It was unnecessary, so i removed it.

rdeits commented on src/Utilities/functions.jl:
```diff
@@ -215,27 +250,52 @@ If `x` (resp. `y`, `z`) is `VariableIndex(1)` (resp. 2, 3).
-        elseif tnot0
-            if !isempty(terms) && iszero(coefficient(last(terms)))
-                terms[end] = t
+canonical(f::Union{SAF, VAF}) = canonicalize!(copy(f))
```
For non-isbits types, a shallow copy should still be appropriate, as we never mutate any terms; we just replace elements in the terms vector. Moreover, `deepcopy` is significantly slower than the `copy` implementation here: 

```julia
julia> f = MOI.ScalarAffineFunction(MOI.ScalarAffineTerm.(rand(1000), MOI.VariableIndex.(rand(1:100, 1000))), 1.0);

julia> using BenchmarkTools

julia> @btime copy($f);
  1.045 μs (2 allocations: 15.78 KiB)

julia> @btime deepcopy($f);
  2.058 μs (5 allocations: 16.22 KiB)
```

rdeits commented on src/Utilities/functions.jl:
```diff
@@ -202,7 +202,42 @@ coefficient(t::MOI.ScalarAffineTerm) = t.coefficient
+Return a new affine function with a shallow copy of the terms and constant(s)
+from `f`.
+"""
+Base.copy(f::F) where {F <: Union{SAF, VAF}} = F(copy(f.terms), copy(_constant(f)))
```
done

mlubin commented on test/runtests.jl:
```diff
@@ -6,6 +6,7 @@ const MOIB = MathOptInterface.Bridges
 
 using Compat
 using Compat.Test
+using Compat.Random
```
Not needed anymore

rdeits replied:
done
